src/main/scala/lms/core/traversal.scala

```diff
@@ -532,14 +535,35 @@ abstract class Transformer extends Traverser {
     // println(s"transformed ${n.n}->${subst(n.n)}")
   }
 
+  var graphCache: Map[Sym, Node] = _
+  var oldSourceMap: mutable.Map[Exp, SourceContext] = _
+
   def transform(graph: Graph): Graph = {
     // XXX unfortunate code duplication, either
     // with traverser or with transform(Block)
+
+    // graphCache may be use during transformation to check the Node of a Sym
+    graphCache = graph.globalDefsCache
+
+    // Handling MetaData 1. save oldTypeMap/SourceMap first
+    val oldTypeMap = Adapter.typeMap
+    oldSourceMap = Adapter.sourceMap
+    // Handling MetaData 2. initialize MetaData as fresh, so the transformer might add new metadata entries
+    Adapter.typeMap = new scala.collection.mutable.HashMap[lms.core.Backend.Exp, Manifest[_]]()
+    Adapter.sourceMap = new scala.collection.mutable.HashMap[lms.core.Backend.Exp, SourceContext]()
+
     val block = g.reify { e =>
       assert(graph.block.in.length == 1)
```

Comment:
what's the case that `block.in.length` is not 1?

Reply:
we do not handle that yet :)